File: ApogeeVGC_CPP/sim/dex-moves/MoveEventMethods.cpp
#include "MoveEventMethods.h"

MoveEventMethods::MoveEventMethods(const MoveEventMethods& other) :
	base_power_callback(other.base_power_callback),
	before_move_callback(other.before_move_callback),
	before_turn_callback(other.before_turn_callback),
	damage_callback(other.damage_callback),
	priority_charge_callback(other.priority_charge_callback),
	on_disable_move(other.on_disable_move),
	on_after_hit(other.on_after_hit),
	on_after_sub_damage(other.on_after_sub_damage),
	on_after_move_secondary_self(other.on_after_move_secondary_self),
	on_after_move_secondary(other.on_after_move_secondary),
	on_after_move(other.on_after_move),
	on_damage_priority(other.on_damage_priority ? std::make_unique<int>(*other.on_damage_priority) : nullptr),
	on_damage(other.on_damage),
	on_base_power(other.on_base_power),
	on_effectiveness(other.on_effectiveness),
	on_hit(other.on_hit),
	on_hit_field(other.on_hit_field),
	on_hit_side(other.on_hit_side),
	on_modify_move(other.on_modify_move),
	on_modify_priority(other.on_modify_priority),
	on_move_fail(other.on_move_fail),
	on_modify_type(other.on_modify_type),
	on_modify_target(other.on_modify_target),
	on_prepare_hit(other.on_prepare_hit),
	on_try(other.on_try),
	on_try_hit(other.on_try_hit),
	on_try_hit_field(other.on_try_hit_field),
	on_try_hit_side(other.on_try_hit_side),
	on_try_immunity(other.on_try_immunity),
	on_try_move(other.on_try_move),
	on_use_move_message(other.on_use_move_message)
{
}

File: ApogeeVGC/Sim/Events/MoveEventMethods.cs
using ApogeeVGC.Sim.BattleClasses;
using ApogeeVGC.Sim.Effects;
using ApogeeVGC.Sim.Moves;
using ApogeeVGC.Sim.PokemonClasses;
using ApogeeVGC.Sim.SideClasses;
using ApogeeVGC.Sim.Utils;

namespace ApogeeVGC.Sim.Events;

public delegate IntFalseUnion? BasePowerCallbackHandler(IBattle battle, Pokemon pokemon, Pokemon target,
    ActiveMove move);

public delegate BoolVoidUnion BeforeMoveCallbackHandler(IBattle battle, Pokemon pokemon, Pokemon? target,
    ActiveMove move);

public delegate void BeforeTurnCallbackHandler(IBattle battle, Pokemon pokemon, Pokemon target, ActiveMove move);
public delegate IntFalseUnion DamageCallbackHandler(IBattle battle, Pokemon pokemon, Pokemon target);
public delegate void PriorityChargeCallbackHandler(IBattle battle, Pokemon pokemon);

public delegate void OnDisableMoveHandler(IBattle battle, Pokemon pokemon);
public delegate void OnAfterSubDamageHandler(IBattle battle, int damage, Pokemon target, Pokemon source,
    ActiveMove move);

public delegate IntBoolVoidUnion? OnDamageHandler(IBattle battle, int damage, Pokemon target, Pokemon source,
    IEffect effect);

public delegate IntVoidUnion OnEffectivenessHandler(IBattle battle, int typeMod, Pokemon? target, PokemonType type,
    ActiveMove move);
public delegate BoolUndefinedVoidUnion? OnHitSideHandler(IBattle battle, Side side, Pokemon source, ActiveMove move);
public delegate void OnModifyMoveHandler(IBattle battle, ActiveMove move, Pokemon pokemon, Pokemon? target);
public delegate void OnModifyTypeHandler(IBattle battle, ActiveMove move, Pokemon pokemon, Pokemon target);

public delegate void OnModifyTargetHandler(IBattle battle, Pokemon relayTarget, Pokemon pokemon, Pokemon target,
    ActiveMove move);

public delegate BoolUndefinedVoidUnion? OnTryHitSideHandler(IBattle battle, Side side, Pokemon source,
    ActiveMove move);

public interface IMoveEventHandlers
{
    TypeUndefinedUnion<BasePowerCallbackHandler>? BasePowerCallback { get; }
    TypeUndefinedUnion<BeforeMoveCallbackHandler>? BeforeMoveCallback { get; }
    TypeUndefinedUnion<BeforeTurnCallbackHandler>? BeforeTurnCallback { get; }
    TypeUndefinedUnion<DamageCallbackHandler>? DamageCallback { get; }
    TypeUndefinedUnion<PriorityChargeCallbackHandler>? PriorityChargeCallback { get; }

    TypeUndefinedUnion<OnDisableMoveHandler>? OnDisableMove { get; }
    TypeUndefinedUnion<VoidSourceMoveHandler>? OnAfterHit { get; }
    TypeUndefinedUnion<OnAfterSubDamageHandler>? OnAfterSubDamage { get; }
    TypeUndefinedUnion<VoidSourceMoveHandler>? OnAfterMoveSecondarySelf { get; }
    TypeUndefinedUnion<VoidMoveHandler>? OnAfterMoveSecondary { get; }
    TypeUndefinedUnion<VoidSourceMoveHandler>? OnAfterMove { get; }
    TypeUndefinedUnion<OnDamageHandler>? OnDamage { get; }
    TypeUndefinedUnion<ModifierSourceMoveHandler>? OnBasePower { get; }
    TypeUndefinedUnion<OnEffectivenessHandler>? OnEffectiveness { get; }
    TypeUndefinedUnion<ResultMoveHandler>? OnHit { get; }
    TypeUndefinedUnion<ResultMoveHandler>? OnHitField { get; }
    TypeUndefinedUnion<OnHitSideHandler>? OnHitSide { get; }
    TypeUndefinedUnion<OnModifyMoveHandler>? OnModifyMove { get; }
    TypeUndefinedUnion<ModifierSourceMoveHandler>? OnModifyPriority { get; }
    TypeUndefinedUnion<VoidMoveHandler>? OnMoveFail { get; }
    TypeUndefinedUnion<OnModifyTypeHandler>? OnModifyType { get; }
    TypeUndefinedUnion<OnModifyTargetHandler>? OnModifyTarget { get; }
    TypeUndefinedUnion<ResultMoveHandler>? OnPrepareHit { get; }
    TypeUndefinedUnion<ResultSourceMoveHandler>? OnTry { get; }
    TypeUndefinedUnion<ExtResultSourceMoveHandler>? OnTryHit { get; }
    TypeUndefinedUnion<ResultMoveHandler>? OnTryHitField { get; }
    TypeUndefinedUnion<OnTryHitSideHandler>? OnTryHitSide { get; }
    TypeUndefinedUnion<ResultMoveHandler>? OnTryImmunity { get; }
    TypeUndefinedUnion<ResultSourceMoveHandler>? OnTryMove { get; }
    TypeUndefinedUnion<VoidSourceMoveHandler>? OnUseMoveMessage { get; }
}